File: src/infinicore/ops/conv2d/conv2d.cc
#include "infinicore/ops/conv2d.hpp"

#include "../../utils.hpp"

#include <stdexcept>

namespace infinicore::op {

common::OpDispatcher<Conv2d::schema> &Conv2d::dispatcher() {
    static common::OpDispatcher<Conv2d::schema> dispatcher_;
    return dispatcher_;
};

void Conv2d::execute(Tensor output,
                     Tensor input,
                     Tensor weight,
                     Tensor bias,
                     const size_t *pads,
                     const size_t *strides,
                     const size_t *dilations,
                     size_t n) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(output, input, weight, bias);
    infinicore::context::setDevice(output->device());
    auto device_type = output->device().getType();
    auto func = dispatcher().lookup(device_type);

    if (func == nullptr) {
        throw std::runtime_error("No Conv2d implementation found for device type: " + std::to_string(static_cast<int>(device_type)));
    }

    func(output, input, weight, bias, pads, strides, dilations, n);
}

Tensor conv2d(Tensor input,
              Tensor weight,
              Tensor bias,
              const std::vector<size_t> &pads,
              const std::vector<size_t> &strides,
              const std::vector<size_t> &dilations) {
    // Output shape should be pre-computed by caller; allocate a conservative placeholder.
    // This helper is rarely used in performance-critical paths.
    Shape shape = input->shape();
    auto output = Tensor::empty(shape, input->dtype(), input->device());
    conv2d_(output, input, weight, bias, pads, strides, dilations);
    return output;
}

void conv2d_(Tensor output,
             Tensor input,
             Tensor weight,
             Tensor bias,
             const std::vector<size_t> &pads,
             const std::vector<size_t> &strides,
             const std::vector<size_t> &dilations) {
    if (pads.size() != strides.size() || pads.size() != dilations.size()) {
        throw std::runtime_error("conv2d_: pads/strides/dilations must have the same size");
    }
    Conv2d::execute(output,
                    input,
                    weight,
                    bias,
                    pads.data(),
                    strides.data(),
                    dilations.data(),
                    pads.size());
}
} // namespace infinicore::op

File: python/infinicore/ops/conv2d.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor, zeros


def _pair(value):
    if isinstance(value, int):
        return [value, value]
    return list(value)


def conv2d(
    input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1, *, out=None
):
    if groups != 1:
        raise NotImplementedError(
            "infinicore.ops.conv2d currently supports groups=1 only"
        )

    stride = _pair(stride)
    padding = _pair(padding)
    dilation = _pair(dilation)
    if bias is None:
        bias = zeros((weight.shape[0],), dtype=weight.dtype, device=weight.device)
    bias_raw = bias._underlying

    if out is None:
        return Tensor(
            _infinicore.conv2d(
                input._underlying,
                weight._underlying,
                bias_raw,
                padding,
                stride,
                dilation,
            )
        )

    _infinicore.conv2d_(
        out._underlying,
        input._underlying,
        weight._underlying,
        bias_raw,
        padding,
        stride,
        dilation,
    )
    return out
